networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/discovery/Eth2DiscoveryService.java

```diff
@@ -0,0 +1,250 @@
+import org.ethereum.beacon.discovery.format.SerializerFactory;
+import org.ethereum.beacon.discovery.scheduler.Schedulers;
+import org.ethereum.beacon.discovery.schema.EnrField;
+import org.ethereum.beacon.discovery.schema.EnrFieldV4;
+import org.ethereum.beacon.discovery.schema.IdentitySchema;
+import org.ethereum.beacon.discovery.schema.IdentitySchemaV4Interpreter;
+import org.ethereum.beacon.discovery.schema.NodeRecord;
+import org.ethereum.beacon.discovery.schema.NodeRecordFactory;
+import org.ethereum.beacon.discovery.storage.NodeBucketStorage;
+import org.ethereum.beacon.discovery.storage.NodeSerializerFactory;
+import org.ethereum.beacon.discovery.storage.NodeTable;
+import org.ethereum.beacon.discovery.storage.NodeTableStorage;
+import org.ethereum.beacon.discovery.storage.NodeTableStorageFactoryImpl;
+import org.ethereum.beacon.discovery.util.Functions;
+import org.javatuples.Pair;
+import tech.pegasys.artemis.networking.p2p.network.P2PNetwork;
+import tech.pegasys.artemis.service.serviceutils.Service.State;
+import tech.pegasys.artemis.service.serviceutils.TypedService;
+import tech.pegasys.artemis.util.async.SafeFuture;
+
+@SuppressWarnings("UnstableApiUsage")
+public class Eth2DiscoveryService extends TypedService<State> implements DiscoveryNetwork {
+
+  private static final Logger logger = LogManager.getLogger(Eth2DiscoveryService.class);
+
+  private static final NodeRecordFactory NODE_RECORD_FACTORY =
+      new NodeRecordFactory(new IdentitySchemaV4Interpreter());
+  private static final SerializerFactory TEST_SERIALIZER =
+      new NodeSerializerFactory(NODE_RECORD_FACTORY);
+
+  private DiscoveryManager dm;
+  private NodeTable nodeTable;
+
+  // core parameters for discovery service
+  private String networkInterface;
+  private int port;
+  private byte[] privateKey; // for generating ENR records
+  private List<String> peers; // for setting boot nodes
+
+  // the network to potentially affect with discovered peers
+  private Optional<P2PNetwork<?>> network = Optional.empty();
+
+  // event bus by which to signal other services
+  private EventBus eventBus;
+
+  @Override
+  public SafeFuture<State> doStart() {
+    dm.start();
+    return SafeFuture.completedFuture(this.getState());
+  }
+
+  @Override
+  public SafeFuture<State> doStop() {
+    dm.stop();
+    return SafeFuture.completedFuture(this.getState());
+  }
+
+  public void setNetwork(P2PNetwork<?> network) {
+    assert network != null;
+    this.network = Optional.of(network);
+  }
+
+  public Optional<P2PNetwork<?>> getNetwork() {
+    return network;
+  }
+
+  public void setEventBus(EventBus eventBus) {
+    assert eventBus != null;
+    this.eventBus = eventBus;
+  }
+
+  public EventBus getEventBus() {
+    return eventBus;
+  }
+
+  @Override
+  public void findPeers() {
+    // nop - is this to start?
+  }
+
+  @Override
+  public void subscribePeerDiscovery(DiscoveryPeerSubscriber subscriber) {
+    eventBus.register(subscriber);
```

Comment:
We have this [`Subscribers` class](https://github.com/PegaSysEng/artemis/blob/ab7ea72efe7eddc72a04d901830e60315489cb59/util/src/main/java/tech/pegasys/artemis/util/events/Subscribers.java#L43) that you could use to track subscribers and dispatch events.  If we're going to keep this API, I'd suggest using that instead of wrapping over the `eventBus`.

Reply:
removed the API